src/optimization/Optimizer.cpp

```diff
@@ -158,82 +93,73 @@ static void generalOptimization(const Module& module, Method& method, const Conf
-const OptimizationPass optimizations::ADD_START_STOP_SEGMENT =
-    OptimizationPass("AddStartStopSegment", addStartStopSegment, 250);
-const OptimizationPass optimizations::EXTEND_BRANCHES = OptimizationPass("ExtendBranches", extendBranches, 300);
-
-const std::set<OptimizationPass> optimizations::DEFAULT_PASSES = {
-    MAP_MEMORY_ACCESS, RUN_SINGLE_STEPS, /* SPILL_LOCALS, */ COMBINE_LITERAL_LOADS, HANDLE_IMMEDIATES,
-    RESOLVE_STACK_ALLOCATIONS, COMBINE_ROTATIONS, GENERAL_OPTIMIZATIONS, ELIMINATE, VECTORIZE, SPLIT_READ_WRITES,
-    REORDER, COMBINE, UNROLL_WORK_GROUPS, ADD_START_STOP_SEGMENT, EXTEND_BRANCHES
-    /* , REMOVE_CONSTANT_LOAD_IN_LOOPS
+Optimizer::Optimizer(const Configuration& config) : config(config)
+{
+    // runs all the single-step optimizations. Combining them results in fewer iterations over the instructions
+    passes.emplace_back("SingleSteps", runSingleSteps);
+    // combines duplicate vector rotations, e.g. introduced by vector-shuffle into a single rotation
+    passes.emplace_back("CombineRotations", combineVectorRotations);
+    passes.emplace_back("GeneralOptimizations", generalOptimization);
+    /*
      * TODO in combination with a bug/missing check in register-allocation, this generates invalid code (e.g.
      * testing/test_barrier.cl) More exact: the load is moved outside the loop but the register is re-assigned in the
      * loop having wrong value for successive iterations In register-allocation, need to check for loops and reserve
      * whole loop
      */
-};
-
-Optimizer::Optimizer(const Configuration& config, const std::set<OptimizationPass>& passes) :
-    config(config), passes(passes)
-{
+    // move constant loads in (nested) loops outside the loops
+    // passes.emplace_back("RemoveConstantLoadInLoops", removeConstantLoadInLoops);
+    // eliminates useless instructions (dead store, move to same, redundant arithmetic operations, ...)
+    passes.emplace_back("EliminateDeadStores", eliminateDeadStore);
+    // vectorizes loops
+    passes.emplace_back("VectorizeLoops", vectorizeLoops);
+    // more like a de-optimization. Splits read-after-writes (except if the local is used only very locally), so the
+    // reordering and register-allocation have an easier job
+    passes.emplace_back("SplitReadAfterWrites", splitReadAfterWrites);
+    // combines loadings of the same literal value within a small range of a basic block
+    passes.emplace_back("CombineLiteralLoads", combineLoadingLiterals);
+    // re-order instructions to eliminate more NOPs and stall cycles
+    passes.emplace_back("ReorderInstructions", reorderWithinBasicBlocks);
+    // run peep-hole optimization to combine ALU-operations
+    passes.emplace_back("CombineALUIinstructions", combineOperations);
+    // add (runtime-configurable) loop over the whole kernel execution, allowing for skipping some of the syscall
+    // overhead for kernels with many work-groups
+    passes.emplace_back("UnrollWorkGroups", unrollWorkGroups);
```

Comment:
`unrollWorkGroups` seem to work like normalization?
It seems that we must not skip it.

Reply:
Good catch, I missed this one:)